### PR TITLE
Emit Batch Status for Bulk Actions

### LINKED ISSUE #1367: Possibility to emit indices in bulk action with `return_stream?: true`
**Is your feature request related to a problem? Please describe.**

I'm writing a Mix task, which imports a large file via a bulk insert. To display progress, I would like to be able to log a message every n entries so that I can see the progress. TO do this, I have to enable `return_records?` at the moment. This uses memory unnecessarily.

**Describe the solution you'd like**

A new option `emit_indices?` which will return the current index instead of the whole record. Only available when `return_stream?: true` and `return_records?: false`.

**Describe alternatives you've considered**

None

**Express the feature either with a change to resource syntax, or with a change to the resource interface**

For example

```elixir
rows
|> Ash.bulk_create(Resource, :action, return_stream?: true, emit_indices?: true)
|> Enum.each(fn
  {:omitted_record, i} when mod(i, 100) -> IO.puts "imported #{i} records"
  _ -> :discard
end)
```

**Additional context**

None

### PR DESCRIPTION
**Pull request by maennchen**
Resolves #1367
Based on #1389 

# Scope

* Introduces the option `return_batch_status?` to all `Ash.bulk_*` operations.
* Adds `record_count` to `Ash.BulkResult`
* Adds `Ash.BulkResult.BatchStatus` and yields it in the result stream when enabled

# TODO

* [x] Support bulk_create
* [ ] Support bulk_update with list
* [ ] Support bulk_update with query
* [ ] Support bulk_destroy with list
* [ ] Support bulk_destroy with query

# Contributor checklist

- [ ] Bug fixes include regression tests
- [ ] Chores
- [ ] Documentation changes
- [ ] Features include unit/acceptance tests
- [ ] Refactoring
- [ ] Update dependencies

### CODE REVIEW

**Comment by maennchen** [comment]
@zachdaniel Is this going in the direction you intended?

File: lib/ash/actions/create/bulk.ex [hunk 1]
```diff
@@ -178,15 +178,17 @@ defmodule Ash.Actions.Create.Bulk do
...
+      |> Stream.concat()
+      |> Ash.BulkResult.BatchStatus.accumulate()
     else
       try do
-        records =
+        {records, record_count} =
           if opts[:return_records?] do
-            Enum.to_list(Stream.concat(changeset_stream))
+            records = Enum.to_list(Stream.concat(changeset_stream))
+            {records, length(records)}
```
Comment (zachdaniel) [inline lib/ash/actions/create/bulk.ex:189]: We should count and convert to  list in a single reduce here I think.

File: lib/ash/actions/create/bulk.ex [hunk 2]
```diff
@@ -1216,12 +1275,38 @@ defmodule Ash.Actions.Create.Bulk do
...
 
             :ok ->
               Process.put({:any_success?, ref}, true)
+
+              cond do
+                opts[:return_records?] ->
+                  raise "`#{inspect(Ash.Resource.Info.data_layer(resource))}.bulk_create/3` returned :ok without a result when `return_records?` is true"
+
+                opts[:return_batch_status?] ->
+                  raise "`#{inspect(Ash.Resource.Info.data_layer(resource))}.bulk_create/3` returned :ok without affected row count when `return_batch_status?` is true"
```
Comment (zachdaniel) [inline lib/ash/actions/create/bulk.ex:1304]: We should synthesize this return type for calls to single creates.

**Review by zachdaniel — DISMISSED** [review]

**Comment by zachdaniel** [comment]
I know this is still WIP but did an initial review. This looks like it's going in the right direction to me 👌

File: lib/ash/actions/create/bulk.ex (hunk 2 above)
Comment (maennchen) [reply lib/ash/actions/create/bulk.ex:1304]: As I understand the current code, it is fully synthesized. The error is only raised if you implement `bulk_create` and return `:ok` from there.

Am I missing something?

That fn could use a refactor, it's quite hard to follow what it does :sweat_smile:

Comment (zachdaniel) [reply lib/ash/actions/create/bulk.ex:1305]: Ah, right because that callback returns `{:ok, result}`.

**Comment by maennchen** [comment]
Closing since I have no time soon to finish this and the attached refactorings required to make it happen.
